security/pkg/nodeagent/test/setup.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// WriteDataToFile writes data into file
+func WriteDataToFile(path string, content string) error {
+	if path == "" {
+		return errors.New("empty file path")
+	}
+	f, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	if _, err = f.WriteString(content); err != nil {
+		return err
+	}
+	_ = f.Sync()
+	return nil
+}
+
+// SetupTest starts Envoy, SDS server, CA server and a dummy backend.
+// The test allow HTTP request flow.
+//                                                     CA server
+//                                                         |
+//                                           +---------SDS server-------+
+//                                           |                          |
+// HTTP request->outbound listener->outbound TLS cluster<-mTLS->inbound TLS listener->inbound cluster->backend
+func SetupTest(t *testing.T, testID uint16) *Env {
+	// Set up credential files for bootstrap config
+	jwtToken := getDataFromFile(istioEnv.IstioSrc+"/security/pkg/nodeagent/test/testdata/jwt-token.jwt", t)
+	if err := WriteDataToFile(proxyTokenPath, jwtToken); err != nil {
+		t.Fatalf("failed to set up token file %s: %v", proxyTokenPath, err)
+	}
+
+	env := &Env{}
+	// Set up test environment for Proxy
+	proxySetup := proxyEnv.NewTestSetup(testID, t)
+	proxySetup.SetNoMixer(true)
+	proxySetup.EnvoyTemplate = getDataFromFile(istioEnv.IstioSrc+"/security/pkg/nodeagent/test/testdata/bootstrap.yaml", t)
+	env.ProxySetup = proxySetup
+	env.OutboundListenerPort = int(proxySetup.Ports().ClientProxyPort)
+
+	env.DumpPortMap(t)
+	ca, err := caserver.NewCAServer(int(proxySetup.Ports().MixerPort))
+	if err != nil {
+		t.Fatalf("failed to start CA server: %+v", err)
+	}
+	env.CAServer = ca
+	env.StartSDSServer(t)
+	return env
+}
+
+// DumpPortMap dumps port allocation status
+// outbound listener      : ClientProxyPort
+// inbound listener       : ServerProxyPort
+// test backend           : BackendPort
+// proxy admin            : AdminPort
+// CSR server             : MixerPort
+func (e *Env) DumpPortMap(t *testing.T) {
+	log.Printf("\n\tport allocation status\t\t\t\n"+
+		"outbound listener\t\t:\t%d\n"+
+		"inbound listener\t\t:\t%d\n"+
+		"test backend\t\t\t:\t%d\n"+
+		"proxy admin\t\t\t:\t%d\n"+
+		"CSR server\t\t\t:\t%d\n", e.ProxySetup.Ports().ClientProxyPort,
+		e.ProxySetup.Ports().ServerProxyPort, e.ProxySetup.Ports().BackendPort,
+		e.ProxySetup.Ports().AdminPort, e.ProxySetup.Ports().MixerPort)
+}
+
+// StartProxy starts proxy.
+func (e *Env) StartProxy(t *testing.T) {
+	if err := e.ProxySetup.SetUp(); err != nil {
+		t.Fatalf("failed to start proxy: %v", err)
+	}
+	log.Println("proxy is running...")
+}
+
+// StartSDSServer starts SDS server
+func (e *Env) StartSDSServer(t *testing.T) {
+	serverOptions := sds.Options{
+		WorkloadUDSPath:   sdsPath,
+		UseLocalJWT:       true,
+		JWTPath:           proxyTokenPath,
+		CAEndpoint:        fmt.Sprintf("127.0.0.1:%d", e.ProxySetup.Ports().MixerPort),
```

Comment:
why mixer?

Reply:
This frame work extends the existing one in istio.io/istio/mixer/test/client/. The test setup allocates a set of ports, and I reuse them for the SDS tests.